File: ml-benchmark/synduce/tree/mpps.f
Inductive Tree = nil Unit | single Int | node {Int, Tree, Tree};

max = \a: Int. \b: Int. if < a b then b else a;

spec = \t: Tree. (fix (
  \f: {Int, Int} -> Tree -> {Int, Int}. \s: {Int, Int}. \t: Tree.
  match t with
    nil _ -> s
  | single a -> {+ s.1 a, max s.2 (+ s.1 a)}
  | node {a, l, r} -> f (f {+ s.1 a, max s.2 (+ s.1 a)} l) r
  end
) {0,0} t).2;

repr = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    nil _ -> nil unit
  | single a -> single a
  | node {a, l, r} -> node {a, f l, f r}
  end
);

main = \t: Tree. spec (repr t);

File: ml-benchmark/synduce/tree/mpps.ml
type tree = Nil of unit | Single of int | Node of int * tree * tree

let max a b = if a < b then b else a

let spec t =
  let rec f s t =
    match t with
    | Nil _ -> s
    | Single a ->
        (match s with
         | (s1, s2) -> (s1 + a, max s2 (s1 + a)))
    | Node (a, l, r) ->
        (match s with
         | (s1, s2) ->
             let s' = (s1 + a, max s2 (s1 + a)) in
             f (f s' l) r)
  in
  match f (0, 0) t with
  | (_, x) -> x

let rec repr t =
  match t with
  | Nil _ -> Nil ()
  | Single a -> Single a
  | Node (a, l, r) -> Node (a, repr l, repr r)

let program t = spec (repr t)